### PR TITLE
Add arithmetic operations +,- between timestamp, timestampz and intervals

### PR DESCRIPTION
**Pull request by mkleen**
## Checklist

 - [ ] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [ ] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: common/src/main/java/io/crate/types/IntervalType.java [hunk 1]
```diff
@@ -141,6 +141,14 @@ public void writeValueTo(StreamOutput out, Period p) throws IOException {
         }
     }
 
+    public static Period add(Period a, Period b) {
```
Comment (mfussenegger) [inline common/src/main/java/io/crate/types/IntervalType.java]: Can we inline those? Shallow methods like this tend to add more complexity due to the indirection. They don't encapsulate anything really.

File: common/src/main/java/io/crate/types/TimestampType.java [hunk 2]
```diff
@@ -151,6 +154,16 @@ static long parseTimestamp(String timestamp) {
         }
     }
 
+    public static long addPeriod(Long timestamp, Period period) {
```
Comment (mfussenegger) [inline common/src/main/java/io/crate/types/TimestampType.java]: I'd inline these methods given that they're very shallow, and the indirection probably causes more complexity than it encapsulates. Or at least move to `sql` where they're used.

File: sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java [hunk 3]
```diff
@@ -160,39 +167,104 @@ public FunctionImplementation getForTypes(List<DataType> args) throws IllegalArg
...
 
+
+
+        @Nullable
+        @Override
+        public List<DataType> getSignature(List<? extends FuncArg> dataTypes) {
+            if (dataTypes.size() == 2) {
+                FuncArg fst = dataTypes.get(0);
+                FuncArg snd = dataTypes.get(1);
+                // TODO: or the other way around
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java]: `interval + timestamp` should also be supported I think?

File: sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java [hunk 4]
```diff
@@ -160,39 +167,104 @@ public FunctionImplementation getForTypes(List<DataType> args) throws IllegalArg
...
+
+        @Override
+        public Long evaluate(TransactionContext txnCtx, Input<Object>... args) {
+            if (args.length != 2) {
+                throw new UnsupportedOperationException("Invalid arguments " + Arrays.toString(args));
+            }
+            Long timestamp = (Long) args[0].value();
+            Period period = (Period) args[1].value();
+            if (timestamp != null && period != null) {
+                switch (operator) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java]: Would be nice if we could do this distinction when the method is created, instead of doing it on runtime per row.

File: sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java [hunk 5]
```diff
@@ -160,39 +167,104 @@ public FunctionImplementation getForTypes(List<DataType> args) throws IllegalArg
...
+                    case "+":
+                        return TimestampType.addPeriod(timestamp, period);
+                    case "-":
+                        return TimestampType.minusPeriod(timestamp, period);
+                    default:
+                        throw new UnsupportedOperationException("Invalid operator on interval " + operator);
+                }
             }
-            return Scalar.withOperator(scalar, operator);
+            throw new IllegalArgumentException("Invalid arguments " + Arrays.toString(args));
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java]: `null` value as args need to result in a `null` value as result:

```
select null + '1 day'::interval;
 ?column? 
----------
 💩
(1 row)

```

File: sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java [hunk 6]
```diff
@@ -160,39 +167,104 @@ public FunctionImplementation getForTypes(List<DataType> args) throws IllegalArg
...
+            DataType fst = dataTypes.get(0);
+            DataType snd = dataTypes.get(1);
+
+            if (snd.id() == IntervalType.ID) {
+                switch (fst.id()) {
+                    case TimestampType.ID_WITH_TZ:
+                    case TimestampType.ID_WITHOUT_TZ:
+                        return new IntervalToTimestampScalar(operator);
+                    default:
+                        throw new UnsupportedOperationException("Invalid operator on intervals");
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java]: More descriptive error message would be great.

File: sql/src/test/java/io/crate/integrationtests/ArithmeticIntegrationTest.java [hunk 7]
```diff
@@ -394,4 +395,73 @@ public void testSelectOrderByScalarOnNullValue() throws Exception {
         execute("select (d - 10) from t order by (d - 10) nulls first limit 2");
         assertThat(response.rows()[0][0], is(nullValue()));
     }
+
+    @Test
+    public void test_add_interval_to_interval() {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/ArithmeticIntegrationTest.java]: I think we should move these to a unit test using `AbstractScalarFunctionsTest`, it'll be at least an order of magnitude faster to run them then.

File: sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java [hunk 8]
```diff
@@ -208,4 +254,89 @@ public static Function of(String name, Symbol first, Symbol second, Set<Function
...
+        private final BiFunction<DateTime, Period, DateTime> operation;
+        private final FunctionInfo info;
+
+        IntervalTimestampScalar(String operator, DataType firstType, DataType secondType, DataType returnType) {
+            this.info = new FunctionInfo(new FunctionIdent("add", Arrays.asList(firstType, secondType)), returnType);
+            switch (operator) {
+                case "+":
+                    operation = DateTime::plus;
+                    break;
+                case "-":
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java]: I missed this earlier, but `interval - timestamp` doesn't make sense. (In PostgreSQL it is also unsupported)

File: sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java [hunk 9]
```diff
@@ -39,19 +43,31 @@
...
+import java.util.Locale;
 import java.util.Set;
+import java.util.function.BiFunction;
 import java.util.function.BinaryOperator;
 
 public class ArithmeticFunctions {
 
     private static final Param ARITHMETIC_TYPE = Param.of(
-        DataTypes.NUMERIC_PRIMITIVE_TYPES, DataTypes.TIMESTAMPZ, DataTypes.TIMESTAMP);
+        DataTypes.NUMERIC_PRIMITIVE_TYPES, DataTypes.TIMESTAMPZ, DataTypes.TIMESTAMP, DataTypes.INTERVAL, DataTypes.UNDEFINED);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java]: Why do we need `UNDEFINED` here?

File: sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java [hunk 10]
```diff
@@ -160,40 +175,96 @@ public FunctionImplementation getForTypes(List<DataType> args) throws IllegalArg
...
         }
+
+        private static boolean isTimestamp(DataType d) {
+            return TIMESTAMP_IDS.contains(d.id());
+        }
+
+        private static final Set<Integer> TIMESTAMP_IDS = Sets.newHashSet(DataTypes.TIMESTAMP.id(),
+                                                                          DataTypes.TIMESTAMPZ.id());
+
+        private static final Set<Integer> NUMERICAL_ID = Sets.newHashSet(DataTypes.DOUBLE.id(),
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java]: I think we've already a definition like this in  `DataTypes` which also includes short, can we use that instead?

File: sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java [hunk 11]
```diff
@@ -208,4 +279,140 @@ public static Function of(String name, Symbol first, Symbol second, Set<Function
...
+                    throw new IllegalArgumentException("Unsupported operator for interval " + operator);
+            }
+        }
+
+        @Override
+        public Period evaluate(TransactionContext txnCtx, Input<Object>... args) {
+            Period period = (Period) args[0].value();
+            Object number = args[1].value();
+
+            BigDecimal millis = new BigDecimal(period.toStandardDuration().getMillis());
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java]: This will fail if the `period` contains months. 

Could we maybe extract this logic here into a separate PR and wrap up the `timestamp +/- interval` cases first?

File: sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java (hunk 9 above)
Comment (mkleen) [reply sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java]: To handle values in the arithmetics e.g.  `null + interval '1 second'`.

File: sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java (hunk 11 above)
Comment (mkleen) [reply sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java]: Good idea.

File: sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java (hunk 9 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java]: This shouldn't be necessary, something like `null::timestamp + interval '1 second'` should work without adding undefined here.

File: sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java [hunk 12]
```diff
@@ -208,4 +263,96 @@ public static Function of(String name, Symbol first, Symbol second, Set<Function
...
     }
+
+    private static final class IntervalArithmeticScalar extends Scalar<Period, Object> {
+
+        private final FunctionInfo info;
+        private final BiFunction<Period, Period, Period> operation;
+
+        IntervalArithmeticScalar(String operator) {
+            this.info = new FunctionInfo(
+                new FunctionIdent("interval", Arrays.asList(DataTypes.INTERVAL, DataTypes.INTERVAL)), DataTypes.INTERVAL);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java]: The `FunctionIdent`  must have the same name that was used to lookup a function

File: sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java [hunk 13]
```diff
@@ -208,4 +263,96 @@ public static Function of(String name, Symbol first, Symbol second, Set<Function
...
+
+        IntervalTimestampScalar(String operator, DataType firstType, DataType secondType, DataType returnType) {
+            this.info = new FunctionInfo(new FunctionIdent("intervaltimestamp", Arrays.asList(firstType, secondType)), returnType);
+            switch (operator) {
+                case "+":
+                    operation = DateTime::plus;
+                    break;
+                case "-":
+                    if (ArithmeticFunctionResolver.isInterval(firstType)) {
+                        throw new IllegalArgumentException("Operation not allowed");
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java]: Can we have a more user friendly error here?

File: sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java [hunk 14]
```diff
@@ -208,4 +263,96 @@ public static Function of(String name, Symbol first, Symbol second, Set<Function
...
+
+            if (fst == null || snd == null) {
+                return null;
+            }
+
+            final Long timestamp;
+            final Period period;
+
+            if (fst instanceof Long && snd instanceof Period) {
+                timestamp = (Long) args[0].value();
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java]: could set the indices in the constructor to avoid the checks at each evaluate call

File: sql/src/test/java/io/crate/expression/scalar/arithmetic/IntervalFunctionTest.java [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+    public void test_interval_to_interval() {
+        assertEvaluate("interval '1 second' + interval '1 second'", Period.seconds(2));
+        assertEvaluate("interval '1100 years' + interval '2000 years'", Period.years(3100));
+        assertEvaluate("interval '-10 years' + interval '1 years'", Period.years(-9));
+        assertEvaluate("interval '2 second' - interval '1 second'", Matchers.is(Period.seconds(1)));
+        assertEvaluate("interval '-1 second' - interval '-1 second'", Matchers.is(Period.seconds(0)));
+        assertEvaluate("interval '1 month' + interval '1 year'", Matchers.is(Period.years(1).withMonths(1)));
+        expectedException.expect(RuntimeException.class);
+        expectedException.expectMessage("Interval field value out of range");
+        assertEvaluate("interval '9223372036854775807'", Matchers.is(Period.seconds(1)));
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/expression/scalar/arithmetic/IntervalFunctionTest.java]: can we please separate error scenarios into separate test cases?

File: sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java (hunk 9 above)
Comment (mkleen) [reply sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java]: Ok, according to psql `null + interval `1 second` is supposed be null, but right now only works if I add `UNDEFINED` here. Is there a better way to handle this case ?

Comment (mfussenegger) [reply sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java]: What is the error message in `null + interval '1 second'` if you remove undefined here?

null/undefined can be casted to anything, so it should resolve to a signature that is more meaningful, e.g. `[timestamp, interval]`

File: sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java [hunk 16]
```diff
@@ -208,4 +263,95 @@ public static Function of(String name, Symbol first, Symbol second, Set<Function
...
+                    };
+            }
+        }
+
+        @Override
+        public Long evaluate(TransactionContext txnCtx, Input<Object>... args) {
+            final Long timestamp;
+            final Period period;
+
+            if (firstValueIsInterval) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java]: I had something like

```
period = (Period) args[periodIdx].value();
timestamp = (Long) args[tsIdx].value();
```

in mind. So that there is no additional branching within `evaluate, other than the `null` checks.

Comment (mkleen) [reply sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java]: 👍

Comment (mkleen) [reply sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java]: `null + interval '1 second'` yields to `io.crate.exceptions.ConversionException: Cannot cast PT1S to type double precision` without an UNDEFINED type.

Comment (mfussenegger) [reply sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java]: Ah okay, let's keep it then, I guess it's because https://github.com/crate/crate/pull/8997/files#diff-d4be2122aeae9ba7a07ccf87d4e864b5R181 won't match then and it tries to cast it to `[double, interval]` for whatever reason.

Comment (mkleen) [reply sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java]: Also `interval '1 second' + null` yields to `Cannot cast PT1S to type [double precision, real, bigint, integer, smallint, char, undefined]` without an INTERVAL type defined. I could cover all these cases in `ArithmeticFunctionResolver.getSignature()`, but what would be the benefit of that ?
